helix-view/src/document.rs

```diff
@@ -692,20 +693,20 @@ impl Document {
         path: &Path,
         encoding: Option<&'static Encoding>,
         config_loader: Option<Arc<syntax::Loader>>,
-        config: Arc<dyn DynAccess<Config>>,
```

Comment:
I'd agree, and I think I have thought about this, but the issue is, that `Diagnostic` needs the language server id (and offset_encoding).

So maybe instead keep using the (non-lsp) `Diagnostic`, do the conversion in `Editor` and make 
 `language_server_id` optional (as you described).
Which would probably require the current conversion stuff below as a method in `Editor` instead, something like `fn doc_diagnostics(&self, &Document) -> impl Iterator<Diagnostic>`.

Reply:
That could work but you will have to add a way to filter these diagnostics (before conversion) parameter. Publish diagnostic only wants to replace diagnostics from one language server (and in #6447 I also do some additional filtering). A closure `filter: impl Fn(&lsp::Diagnostic) -> bool` that you then pass to `Iterator::filter` would work well I think

Reply:
I've added the `doc_diagnostics` method. It got a little bit bigger and pyramidic than anticipated though... And I couldn't easily use it in `application.rs` due to borrow-checking issues, not sure whether it makes sense to work around them just for this function.

Reply:
I actually added work around for the borrow checker issues in #6447. It's not too hard since everything relevant is an Arc so you can simply clone everything (mostly document text and config IIRC)

Reply:
Yes I know (as I tried that), it's dependent only on `text`, `path` and because we want to filter for enabled diagnostics `language_config`, but I think it's more convoluted that way and it should also be a little bit less efficient (not that that matters with these comparatively small `n`).